stuffer/s2n_stuffer_pem.c

```diff
@@ -31,82 +31,76 @@
 #define S2N_PEM_DH_PARAMETERS               "DH PARAMETERS"
 #define S2N_PEM_CERTIFICATE                 "CERTIFICATE"
 
-static int s2n_stuffer_data_from_pem(struct s2n_stuffer *pem, struct s2n_stuffer *asn1, const char *keyword)
-{
-    uint8_t linepad[S2N_PEM_LINE_LENGTH + 2];
-    struct s2n_blob line_blob = { .data = linepad, .size = S2N_PEM_LINE_LENGTH + 2 };
-    struct s2n_stuffer line;
-    uint8_t *field;
-
-    GUARD(s2n_stuffer_init(&line, &line_blob));
-
-    /* Skip all data before the start of the PEM headers */
-    GUARD(s2n_stuffer_skip_to_char(pem, '-'));
-
-    /* Check that the first none whitespace line matches the header */
-    GUARD(s2n_stuffer_read_token(pem, &line, '\n'));
-    field = s2n_stuffer_raw_read(&line, sizeof(S2N_PEM_BEGIN_TOKEN) - 1);
-    notnull_check(field);
-    if (memcmp(field, S2N_PEM_BEGIN_TOKEN, sizeof(S2N_PEM_BEGIN_TOKEN) - 1)) {
-        S2N_ERROR(S2N_ERR_INVALID_PEM);
-    }
+static int s2n_stuffer_pem_read_encapsulation_line(struct s2n_stuffer *pem, const char* encap_marker, const char *keyword) {
+    /* Skip any number of Chars until "-----BEGIN " or "-----END " is reached */
+    GUARD(s2n_stuffer_skip_read_until(pem, encap_marker));
 
-    field = s2n_stuffer_raw_read(&line, strlen(keyword));
-    notnull_check(field);
-    if (memcmp(field, keyword, strlen(keyword))) {
-        S2N_ERROR(S2N_ERR_INVALID_PEM);
-    }
+    /* Ensure the keyword matches what's expected (Eg "CERTIFICATE", "PRIVATE KEY", etc) */
+    GUARD(s2n_stuffer_read_expected_str(pem, keyword));
 
-    field = s2n_stuffer_raw_read(&line, sizeof(S2N_PEM_LINE) - 1);
-    notnull_check(field);
-    if (memcmp(field, S2N_PEM_LINE, sizeof(S2N_PEM_LINE) - 1)) {
-        S2N_ERROR(S2N_ERR_INVALID_PEM);
-    }
+    /* Ensure 5 dashes after keyword */
+    GUARD(s2n_stuffer_read_expected_str(pem, S2N_PEM_LINE));
+
+    /* Skip newlines and other whitepsace that may be after the dashes */
+    GUARD(s2n_stuffer_skip_whitespace(pem));
+    return 0;
+}
+
+static int s2n_stuffer_pem_read_begin(struct s2n_stuffer *pem, const char *keyword)
+{
+    return s2n_stuffer_pem_read_encapsulation_line(pem, S2N_PEM_BEGIN_TOKEN, keyword);
+}
 
-    /* Get the actual base64 data */
-    do {
-        GUARD(s2n_stuffer_rewrite(&line));
+static int s2n_stuffer_pem_read_end(struct s2n_stuffer *pem, const char *keyword)
+{
+    return s2n_stuffer_pem_read_encapsulation_line(pem, S2N_PEM_END_TOKEN, keyword);
+}
 
-        /* Per RFC7468 Section 2: PEM parsers must handle different newline conventions.
-         * Support both LF and CR+LF.
-         */
-        GUARD(s2n_stuffer_read_line(pem, &line));
+static int s2n_stuffer_pem_read_contents(struct s2n_stuffer *pem, struct s2n_stuffer *asn1)
+{
+    uint8_t linepad[64] = { 0 };
```

Comment:
*Unrelated to this specific line* : Should we remove the `S2N_PEM_LINE_LENGTH` macro? It doesn't appear to be used with this new implementation.

Reply:
Suggest using a different name than `linepad` here since nothing in this function is line-oriented. Maybe `base64_buf` -> `base64_blob` -> `base64_stuffer`?